hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java

```diff
@@ -547,6 +562,39 @@ protected void processContainer(ContainerInfo container,
     }
   }
 
+  /**
+   * Sends a replicate command for each replica specified in
+   * vulnerableUnhealthy.
+   * @param replicaCount RatisContainerReplicaCount for this container
+   * @param vulnerableUnhealthy List of UNHEALTHY replicas that need to be
+   * replicated
+   * @param placementStatus placement status
+   */
+  private void handleVulnerableUnhealthyReplicas(
+      RatisContainerReplicaCount replicaCount,
+      List<ContainerReplica> vulnerableUnhealthy,
+      ContainerPlacementStatus placementStatus) {
+    ContainerInfo container = replicaCount.getContainer();
+    LOG.debug("Handling vulnerable UNHEALTHY replicas {} for container {}.",
+        vulnerableUnhealthy, container);
+    int pendingAdds = getInflightAdd(container.containerID());
+    if (pendingAdds >= vulnerableUnhealthy.size()) {
+      LOG.debug("There are {} pending adds for container {}, while " +
+              "the number of UNHEALTHY replicas is {}.", pendingAdds,
+          container.containerID(), vulnerableUnhealthy.size());
+      return;
+    }
```

Comment:
I'm not sure about this part. Ideally, we should avoid sending a replicate command for a replica that already has a pending add. Any suggestions on how we can achieve this with pull replication in legacy?

Reply:
I don't think the current structure allows this. I just contains the containerID and the target DN. All you can really do is wait until there are no pending adds I guess. Either they will have completed or timed out.

Reply:
Ok, I've kept the code as is then.